### PR TITLE
Update hand-written resources with `AtLeastOneOf` and `ExactlyOneOf` attributes

### PR DESCRIPTION
**Pull request by megan07**
Due to the known issue that nested blocks that have max items more than one, there are a handful of cases where the new attributes won't work, and thus I created CustomizeDiff blocks for them.

I will create doc changes in future PRs associated with this, but wanted to break up the PRs a little (even though this one still seems a little big).

**Release Note Template for Downstream PRs (will be copied)**

```release-note:breaking-change
`app_engine`: `ssl_management_type` is now required on `google_app_engine_domain_mapping.ssl_settings`
```

```release-note:breaking-change
`app_engine`: `shell` is now required on `google_app_engine_standard_app_version.entrypoint`.
```

```release-note:breaking-change
`app_engine`: `source_url` is now required on `google_app_engine_standard_app_version.deployment.files` and `google_app_engine_standard_app_version.deployment.zip`.
```

```release-note:breaking-change
`app_engine`: `split_health_checks ` is now required on `google_app_engine_application.feature_settings`
```

```release-note:breaking-change
`bigtable`: `cluster_id` is now required on `google_bigtable_app_profile.single_cluster_routing`.
```

```release-note:breaking-change
`bigquery`: At least one of `range` or `skip_leading_rows` are now required on `google_bigquery_table.external_data_configuration.google_sheets_options`.
```

```release-note:breaking-change
`cloudbuild`: `branch` is now required on `google_cloudbuild_trigger.github.branch`.
```

```release-note:breaking-change
`cloudbuild`: `steps` is now required on `google_cloudbuild_trigger.build`.
```

```release-note:breaking-change
`cloudbuild`: `name` is now required on `google_cloudbuild_trigger.build.steps`.
```

```release-note:breaking-change
`cloudbuild`: `name` and `path` are now required on `google_cloudbuild_trigger.build.steps.volumes`.
```

```release-note:breaking-change
`cloutiot`: `public_key_certificate` is now required on `google_cloudiot_registry. credentials `.
```

```release-note:breaking-change
`cloudrun`: `template` is now required on `google_cloudrun_service.spec`.
```

```release-note:breaking-change
`cloudscheduler`: `service_account_email` is now required on `google_cloudscheduler_job.http_target.oauth_token` and `google_cloudscheduler_job.http_target.oidc_token`.
```

```release-note:breaking-change
`composer`: At least one of `airflow_config_overrides`, `pypi_packages`, `env_variables, `image_version`,  or `python_version` are now required on `google_composer_environment.config.software_config`.
```

```release-note:breaking-change
`composer`: `use_ip_aliases` is now required on `google_composer_environment.config.node_config.ip_allocation_policy`.
```

```release-note:breaking-change
`composer`: `enable_private_endpoint` is now required on `google_composer_environment.config.private_environment_config`.
```

```release-note:breaking-change
`compute`: `raw_key` is now required on `google_compute_snapshot.snapshot_encryption_key`.
```

```release-note:breaking-change
`compute`: At least one of `auto_delete`, `device_name`, `disk_encryption_key_raw`, `kms_key_self_link`, `initialize_params`, `mode` or `source` are now required on `google_compute_instance.boot_disk`.
```

```release-note:breaking-change
`compute`: At least one of `size`, `type`, `image`, or `labels` are now required on `google_compute_instance.boot_disk.initialize_params`.
```

```release-note:breaking-change
`compute`: At least one of `network` or `subnetwork` are now required on `google_compute_instance.network_interface`.
```

```release-note:breaking-change
`compute`: At least one of `nat_ip`, `network_tier`, or `public_ptr_domain_name` are now required on `google_compute_instance.network_interface.access_config`.
```

```release-note:breaking-change
`compute`: At least one of `enable_secure_boot`, `enable_vtpm`, or `enable_integrity_monitoring` are now required on `google_compute_instance.shielded_instance_config`.
```

```release-note:breaking-change
`compute`: At least one of `on_host_maintenance`, `automatic_restart`, `preemptible`, or `node_affinities` are now required on `google_compute_instance.scheduling`.
```

```release-note:breaking-change
`compute`: `interface` is now required on `google_compute_instance.scratch_disk`.
```

```release-note:breaking-change
`compute`: Exactly one of `fixed` or `percent` are now required on `google_compute_instance_group_manager.version.target_size`.
```

```release-note:breaking-change
`compute`: At least one of `enable_secure_boot`, `enable_vtpm`, or `enable_integrity_monitoring` are now required on `google_compute_instance_template.shielded_instance_config`.
```

```release-note:breaking-change
`compute`: At least one of `on_host_maintenance`, `automatic_restart`, `preemptible`, or `node_affinities` are now required on `google_compute_instance_template.scheduling`.
```

```release-note:breaking-change
`compute`: At least one of `network` or `subnetwork` are now required on `google_compute_instance_template.network_interface`.
```

```release-note:breaking-change
`compute`: At least one of `nat_ip` or `network_tier` are now required on `google_compute_instance_template.network_interface.access_config`.
```

```release-note:breaking-change
`compute`: At least one of `source` or `source_image` is now required on `google_compute_instance_template.disk`.
```

```release-note:breaking-change
`compute`: `kms_key_self_link` is now required on `google_compute_instance_template.disk.disk_encryption_key`.
```

```release-note:breaking-change
`compute`: Exactly one of `fixed` or `percent` is now required on `google_compute_region_instance_group_manager.version.target_size`.
```

```release-note:breaking-change
`compute`:  `range` is now required on `google_compute_router_peer. advertised_ip_ranges`.
```

```release-note:breaking-change
`container`:  `disabled` is now required on `google_container_cluster.addons_config.http_load_balancing`, `google_container_cluster.addons_config.horizontal_pod_autoscaling`, `google_container_cluster.addons_config.network_policy_config`, `google_container_cluster.addons_config.cloudrun_config`, and `google_container_cluster.addons_config.istio_config`.
```

```release-note:breaking-change
`container`: At least one of `http_load_balancing`, `horizontal_pod_autoscaling` , `network_policy_config`, `cloudrun_config`, or `istio_config` is now required on `google_container_cluster.addons_config`.
```

```release-note:breaking-change
`container`:  `enabled` is now required on `google_container_cluster.network_policy`.
```

```release-note:breaking-change
`container`:  `enable_private_endpoint` is now required on `google_container_cluster.private_cluster_config`.
```

```release-note:breaking-change
`container`:  `enabled` is now required on `google_container_cluster.vertical_pod_autoscaling`.
```

```release-note:breaking-change
`container`:  `cidr_blocks` is now required on `google_container_cluster.master_authorized_networks_config`.
```

```release-note:breaking-change
`compute`:  `channel` is now required on `google_container_cluster.relase_channel`.
```

```release-note:breaking-change
`container`: At least one of `username`, `password` or `client_certificate_config` is now required on `google_container_cluster.master_auth`.
```

```release-note:breaking-change
`dataproc`: At least one of `staging_bucket`, `gce_cluster_config`, `master_config`, `worker_config`, `preemptible_worker_config`, `software_config`, `initialization_action` or `encryption_config` is now required on `google_dataproc_cluster.cluster_config`.
```

```release-note:breaking-change
`dataproc`: At least one of `zone`, `network`, `subnetwork`, `tags`, `service_account`, `service_account_scopes`, `internal_ip_only` or `metadata` is now required on `google_dataproc_cluster.cluster_config.gce_cluster_config`.
```

```release-note:breaking-change
`dataproc`: At least one of `num_instances`, `image_uri`, `machine_type`, `min_cpu_platform`, `disk_config`, or `accelerators` is now required on `google_dataproc_cluster.cluster_config.master_config` and `google_dataproc_cluster.cluster_config.worker_config`.
```

```release-note:breaking-change
`dataproc`: At least one of `num_local_ssds`, `boot_disk_size_gb` or `boot_disk_type` is now required on `google_dataproc_cluster.cluster_config.preemptible_worker_config.disk_config`, `google_dataproc_cluster.cluster_config.master_config.disk_config` and `google_dataproc_cluster.cluster_config.worker_config.disk_config`.
```

```release-note:breaking-change
`dataproc`: At least one of `num_instances` or `disk_config` is now required on `google_dataproc_cluster.cluster_config.preemptible_worker_config`.
```

```release-note:breaking-change
`dataproc`: At least one of `image_version`, `override_properties` or `optional_components` is now required on `google_dataproc_cluster.cluster_config.software_config`.
```

```release-note:breaking-change
`dataproc`: `policy_uri` is now required on `google_dataproc_cluster.cluster_config.autoscaling_config`.
```

```release-note:breaking-change
`dataproc`: `max_failures_per_hour` is now required on `google_dataproc_job.scheduling`.
```

```release-note:breaking-change
`dataproc`: `driver_log_levels` is now required on `google_dataproc_job.pyspark_config.logging_config`, `google_dataproc_job.spark_config.logging_config`, `google_dataproc_job.hadoop_config.logging_config`, `google_dataproc_job.hive_config.logging_config`, `google_dataproc_job.pig_config.logging_config`, `google_dataproc_job.sparksql_config.logging_config`.
```

```release-note:breaking-change
`dataproc`: At least one of `main_class` or `main_jar_file_uri` is now required on `google_dataproc_job.spark_config` and `google_dataproc_job.hadoop_config`.
```

```release-note:breaking-change
`dataproc`: At least one of `query_file_uri` or `query_list` is now required on `google_dataproc_job.hive_config`, `google_dataproc_job.pig_config`, and `google_dataproc_job.sparksql_config`.
```

```release-note:breaking-change
`dns`: `networks` is now required on `google_dns_managed_zone.private_visibility_config`.
```

```release-note:breaking-change
`dns`: `network_url` is now required on `google_dns_managed_zone.private_visibility_config.networks`.
```

```release-note:breaking-change
`iam`: Exactly one of `list_policy`, `boolean_policy`, or `restore_policy` is now required on `google_organization_policy`.
```

```release-note:breaking-change
`iam`: Exactly one of `all` or `values` is now required on `google_organization_policy.list_policy.allow` and `google_organization_policy.list_policy.deny`.
```

```release-note:breaking-change
`monitoring`: `content` is now required on `google_monitoring_uptime_check_config.content_matchers`.
```

```release-note:breaking-change
`sql`: `name` and `value` are now required on `google_sql_database_instance.settings.database_flags`.
```

```release-note:breaking-change
`sql`: At least one of `binary_log_enabled`, `enabled`, `start_time`, and `location` is now required on `google_sql_database_instance.settings.backup_configuration`.
```

```release-note:breaking-change
`sql`: At least one of `authorized_networks`, `ipv4_enabled`, `require_ssl`, and `private_network` is now required on `google_sql_database_instance.settings.ip_configuration`.
```

```release-note:breaking-change
`sql`: At least one of `day`, `hour`, and `update_track` is now required on `google_sql_database_instance.settings.maintenance_window`.
```

```release-note:breaking-change
`sql`: At least one of `cert`, `common_name`, `create_time`, `expiration_time`, and `sha1_fingerprint` is now required on `google_sql_database_instance.settings.server_ca_cert`.
```

```release-note:breaking-change
`sql`: At least one of `ca_certificate`, `client_certificate`, `client_key`, `connect_retry_interval`, `dump_file_path`, `failover_target`, `master_heartbeat_period`, `password`, `ssl_cipher`, `username`, and `verify_server_certificate` is now required on `google_sql_database_instance.settings.replica_configuration`.
```

```release-note:breaking-change
`sql`: `value` is now required on `google_sql_database_instance.settings.ip_configuration.authorized_networks`.
```

```release-note:breaking-change
`storage`: At least one of `main_page_suffix` or `not_found_page` is now required on `google_storage_bucket.website`.
```

```release-note:breaking-change
`storage`: At least one of `origin`, `method`, `response_header`, or `max_age_seconds` is now required on `google_storage_bucket.cors`.
```

```release-note:breaking-change
`storage`: At least one of `min_time_elapsed_since_last_modification`, `max_time_elapsed_since_last_modification`, `include_prefixes`, and `exclude_prefixes` is now required on `google_storage_transfer_job.transfer_spec.object_conditions`.
```

```release-note:breaking-change
`storage`: At least one of `overwrite_objects_already_existing_in_sink`, `delete_objects_unique_in_sink`, and `delete_objects_from_source_after_transfer` is now required on `google_storage_transfer_job.transfer_spec.transfer_options`.
```

```release-note:breaking-change
`storage`: At least one of `gcs_data_source`, `aws_s3_data_source`, and `http_data_source` is now required on `google_storage_transfer_job.transfer_options`.
```

### CODE REVIEW

File: third_party/terraform/tests/resource_dataproc_job_test.go [hunk 1]
```diff
@@ -22,21 +22,22 @@ type jobTestField struct {
...
-		CheckDestroy: testAccCheckDataprocJobDestroy,
-		Steps: []resource.TestStep{
-			{
-				Config:      testAccDataprocJob_missingJobConf(),
-				ExpectError: regexp.MustCompile("You must define and configure exactly one xxx_config block"),
-			},
-		},
-	})
-}
+// TODO (mbang): Test `ExactlyOneOf` here
```
Comment (megan07) [inline third_party/terraform/tests/resource_dataproc_job_test.go]: With using `ExactlyOneOf`, each field will produce an error and it seemed, last I tested this, that the order of those errors wasn't consistent, so I need to look back into how to check this.

File: third_party/terraform/website/docs/guides/version_3_upgrade.html.markdown [hunk 2]
```diff
@@ -53,19 +53,30 @@ so Terraform knows to manage them.
```
Comment (megan07) [inline third_party/terraform/website/docs/guides/version_3_upgrade.html.markdown]: I don't think I added every change in here, but I plan to go through all the docs and update them.

**Comment by ndmckinley** [comment]
This looks great.  Do you have the test results on this?  I wonder if any of our unit tests will fail because we are missing a required block.

**Comment by megan07** [comment]
@ndmckinley so far I have run a subset of tests.  anything that I've added the `AtLeastOneOf` or `ExactlyOneOf` flag to, I've at least ran one of the tests for, and they've all seemed to pass.  I'd love to run all the tests (if not now, eventually - maybe over the weekend?), to make sure nothing in a different test that's using these resources is broken, I just haven't had that time yet.

**Pull request merged**